File: app/assets/javascripts/climb.js
var Climb = function(attr) {
	this.id = attr.id;
	this.name = attr.name;
	this.category = attr.category;
	this.rating = attr.rating;
	this.integer_rating = parseInt(attr.integer_rating);
	this.height = parseInt(attr.height);
	this.pitches = parseInt(attr.pitches);
	this.url = attr.url;
	this.state = attr.state_id || "--";
	this.area = attr.area_id || "--";
	this.subarea = attr.subarea_id || "--";
	this.wall = attr.wall_id || "--";
	this.crag = attr.crag_id || "--";
	this.section = attr.section_id || "--";
	this.face = attr.face_id || "--";
};

Climb.prototype.locationString = function() {
	var locations = [this.state, this.area, this.subarea, this.crag, this.section, this.face],
			string = "";
	locations.forEach(function(obj, index, locations){
		if ( obj === "--") {
			return;
		}
		else if (index === 0) {
			string += obj.replace(/"|'/g, '&apos;');;
		} else if (obj) {
			string += " > " + obj.replace(/"|'/g, '&apos;');;
		} else {
			return;
		};
	});
	return string;
};

Climb.prototype.searchResultView = function(val) {
	var neededShit = {category: this.category, rating: this.rating, height: this.height, pitches: this.pitches},
			html_string = "<li data-id='"+this.id+"' class='routeLine'><p class='name inline-block center-text ui-draggable-handle'>"+this.listName()+"</p>",
			tooltip = this.locationString();
	for (var key in neededShit) {
		if (neededShit[key]) {
			html_string += "<p class='"+key+" inline-block' >"+neededShit[key]+"</p>";
		} else {
			html_string += "<p class='"+key+" inline-block'>--</p>";
		}

	}
	return html_string += "<p class='"+this.locationAttrView(val)+" inline-block' title='"+tooltip+"' data-toggle='tooltip' data-placement='bottom'>"+this[this.locationAttrView(val)]+"</p><p class='wall inline-block'>"+this.listWall()+"</p><p class='inline-block float-right link-icon'><a href='"+this.url+"', target='_blank'><i class='fa fa-external-link'></i></a></p></li>";
};

Climb.prototype.tickListView = function() {
	var html_string = "<li data-id='"+this.id+"' class='tickListItem'><p class='inline-block route-name center-text'>"+this.name+"</p>";
	html_string += "<div class='tick-list-sub-line'><span class='category'>"+this.category+"</span>";
	html_string += "<span class='rating'>"+this.rating+"</span>";
	return html_string += "<span class='state'>"+this.state+"</span><i class='fa fa-trash float-right delete-climb pointer'></i></div></li>"
};

Climb.prototype.locationAttrView = function(val) {
	location_things = ["area",	"subarea", "crag",	"section",	"face"]
	return location_things.indexOf(val) > -1 ? val : "state"
};

Climb.prototype.listName = function() {
	if (this.name.length > 17) {
		return this.name.slice(0, 17) + "...";
	} else {
		return this.name;
	};
};

Climb.prototype.listWall = function() {
	if (this.wall){
		if (this.wall.length > 17) {
			return this.wall.slice(0, 15) + "..."
		} else {
			return this.wall;
		};
	};
	return;
};

$(document).ready(function(){
	$("body").on("click", ".delete-climb", function(event){
		var id = $(event.target).parent().parent().attr("data-id"),
		url = $("body").find("#tickListContainer ol").attr("data-url") + "/delete_climb/" + id,
		par = $(event.target).parent().parent();
		$.get(url, function(){
			par.hide(200, function(){
				par.remove()
			});
			;
		});
	});
});

File: app/models/climb.rb
class Climb < ActiveRecord::Base

  has_many :appearances
  has_many :tick_lists, through: :appearances

  belongs_to :state
  belongs_to :area
  belongs_to :subarea
  belongs_to :crag
  belongs_to :section
  belongs_to :face
  belongs_to :wall
  serialize  :tags, Array

  searchable do
  	text :name

  	integer :height, :pitches, :integer_rating
  	text :state do
      state.name unless state.nil?
    end
  	text :area do
      area.name  unless area.nil?
    end
  	text :subarea do
      subarea.name  unless subarea.nil?
    end
  	text :crag do
      crag.name  unless crag.nil?
    end
  	text :section do
      section.name  unless section.nil?
    end
    text :face do
      face.name  unless face.nil?
    end
  	text :wall do
      wall.name  unless wall.nil?
    end
  end
  
  def to_info
    unneeded_shit = ["created_at", "updated_at", "tags"]
    info = attributes
    info["state_id"] = state.name unless state.nil?
    info["area_id"] = area.name unless area.nil?
    info["subarea_id"] = subarea.name unless subarea.nil?
    info["crag_id"] = crag.name unless crag.nil?
    info["section_id"] = section.name unless section.nil?
    info["face_id"] = face.name unless face.nil?
    info["wall_id"] = wall.name unless wall.nil?
    info.each do |k,v| 
      info.delete(k) if unneeded_shit.include?(k) || v.nil?
    end
    return info
  end

  def to_html
    html_string = "<li data-id='#{id}' class='routeLine'>"
    html_string << "<p class='inline-block route-name center-text'>#{name}</p>"
    html_string << "<p class='inline-block'>#{category}</p>"
    html_string << "<p class='inline-block'>#{rating}</p>"
    html_string << "<p class='inline-block'>#{height}</p>"
    html_string << "<p class='inline-block'>#{pitches}</p>"
    html_string << "<p class='inline-block'><a href='#{url}', target='_blank'><i class='fa fa-external-link'></i></a></p>"
    return html_string << "</li>"
  end

  def to_tick_list
    html_string = "<li data-id='#{id}' class='tickListItem'>"
    html_string << "<p class='inline-block route-name center-text'>#{name}</p>"
    html_string << "<div class='tick-list-sub-line'><span class='category'>#{category}</span>"
    html_string << "<span class='rating'>#{rating}</span>"
    html_string << "<span class='state'>#{state.name}</span>"
    html_string << "<i class='fa fa-trash float-right delete-climb pointer'></i></div>"
    return html_string << "</li>"
  end

  def self.query(string)
  	query = self.search do
  		fulltext string
  	end
  	return query.results.map(&:id)
  end

end

